### ISSUE TITLE
is there a way to know the size of a git-annex folder

### ISSUE THREAD

**Issue by jcohenadad** [neuropoly/data-management#51]
e.g. `sct-testing-large` takes almost 40GB on my drive. I had to do emergency cleaning before being able to review `af/fix_karoTobiasMS`.

**Comment by kousu** [comment]
You can do:

```
du -hs .git/objects
du -hs .git/annex/objects
```

to compare the distribution of space between regular git and the annex. I haven't seen a command that can *predict* how large an annex is before downloading it.

In theory it should be possible, since each annexed file encodes its size, e.g.:

```
[kousu@requiem data-multi-subject]$ git show HEAD:sub-cmrra01/anat/sub-cmrra01_acq-MToff_MTS.nii.gz 
/annex/objects/SHA256E-s1755486--e3fe6b9a171e62f219b75893f1e358c513594b3801e3f9afe13f84bcd7ba3aca.nii.gz
```

is 1755486 bytes, or 1.7MB. So if you have cloned a repo but not run `sync --content` yet in theory you can count up what you're in for.

I think I want to recommend everyone doing:

```
git config --global annex.thin=true
```

which will prevent duplicating files in the annex when they are checked out -- it will use hardlinks, aka filesystem pointers, instead. But this conflicts with `annex.hardlink=true` which is a different, older, feature that we [might want to use](https://git-annex.branchable.com/tips/local_caching_of_annexed_files/) on processing servers to let multiple people share a repo. So I don't know.

**Comment by jcohenadad** [comment]
> By the way, I think I want to recommend everyone doing:
git config --global annex.thin=true
which will prevent duplicating files in the annex when they are checked out -- it will use hardlinks, aka filesystem pointers, instead. But this conflicts with annex.hardlink=true which is a different, older, feature that we might want to use on processing servers to let multiple people share a repo. So I don't know.

i see more pros than cons (ie: 95% of our usage will be one person using one repos on their local home/scratch)

**Comment by kousu** [comment]
We should integrate this into https://github.com/neuropoly/management/issues/49

**Comment by jcohenadad** [comment]
@kousu should i run this command in a git annex repos?
```
julien-macbook:~/data/data.neuro $ git config --global annex.thin=true
error: invalid key: annex.thin=true
```

**Comment by kousu** [comment]
No `=`:

`git config --global annex.thin true`

**Comment by kousu** [comment]
I've already made a custom gitolite command in https://github.com/kousu/gitolite-mods/; we can add one that just runs `du -s` for each repo. It won't be totally accurate but it'll be good enough? It'll require some custom code though.

**Comment by jcohenadad** [comment]
could we consider a script that would run as root (but that could be launched by neuropoly users), which would get the size of each repos and report it on the user's terminal?

**Comment by kousu** [comment]
> I've already made a custom gitolite command in https://github.com/kousu/gitolite-mods/; we can add one that just runs `du -s` for each repo. It won't be totally accurate but it'll be good enough? It'll require some custom code though.

It won't take long but I'm not going to shift gears into perl to write it right now. Getting it right could be fiddly too.

Alternately: this is another motivation for #68. `git` can't tell you in advance of downloading but it can tell you as it downloads, which is almost as good.

**Comment by jcohenadad** [comment]
> Alternately: this is another motivation for #68. git can't tell you in advance of downloading but it can tell you as it downloads, which is almost as good.

yes indeed. Can we start a prototype with a dataset, and see if we are happy with it?

**Comment by jcohenadad** [comment]
@kousu I followed instructions to use `git config --global annex.thin true`, but still ended up with a folder double the size:

On my laptop after doing this:
~~~
git config --global annex.thin true
git clone git@data.neuro.polymtl.ca:datasets/msseg_challenge_2021
cd msseg_challenge_2021
git annex init
git annex sync --content
cd ..
du -s msseg_challenge_2021/
11552288	msseg_challenge_2021/
~~~

On duke (non-git annexed dataset):
~~~
julien-macbook:/Volumes/projects/ivadomed $ du -sh ms_challenge_2021_bids_DoNotUse/
5.3G	ms_challenge_2021_bids_DoNotUse/
~~~

So, the dataset on my local laptop is double the size.

Note: I'm a bit confused with the use of "du". On my local folder, when I use `-sh` I get only 5.5G
~~~
julien-macbook:~/data/data.neuro $ du -sh msseg_challenge_2021/
5.5G	msseg_challenge_2021/
~~~
However I _know_ this is wrong, because when I use a software like "Disk Space Analyzer Pro" it tells me the total folder size is 11GB. Also, when _not_ using the flag "h", I get: `11552288`.

Another thing I do not understand, is that when running `du -s` on the remote folder (duke), I get double the size???
~~~
du -s /Volumes/projects/ivadomed/ms_challenge_2021_bids_DoNotUse 
11085280	/Volumes/projects/ivadomed/ms_challenge_2021_bids_DoNotUse
~~~

What's going on???

**Comment by kousu** [comment]
@jcohenadad the GUI app probably doesn't understand hardlinks. If you run `du -s -l` I bet it will report the same numbers (`-l` disables the counter's hardlink deduplication). Similarly, if you were to look at `df` or at Disk Utility before and after you would see you're only using 5GB.

I'm not sure what the problem with the `smb://duke` copy is; *I* see only one time the size:

```
u108545@joplin:~/duke/projects/ivadomed/ms_challenge_2021_bids_DoNotUse$ du -s .
5542642	.
```

And it doesn't seem to have anything to do with hardlinks:

```
u108545@joplin:~/duke/projects/ivadomed/ms_challenge_2021_bids_DoNotUse$ du -s -l
5542642	.
u108545@joplin:~/duke/projects/ivadomed/ms_challenge_2021_bids_DoNotUse$ du -s
5542642	.
```

Maybe something odd about the Mac smb:// driver? Try...rebooting?

**Comment by kousu** [comment]
Oh hang on, is this an accurate mac manpage: https://ss64.com/osx/du.html. It says 

> Files having multiple hard links are counted (and displayed) a single time per du execution. 

and doesn't give a `-l` to turn it off.

There's another way to check: pick a file and run `stat`; you'll see it has "Links: 2":

```
nguenther@data:~/datasets/msseg_challenge_2016$ stat sub-01016SACH/anat/sub-01016SACH_ce-T1w.nii.gz 
  File: sub-01016SACH/anat/sub-01016SACH_ce-T1w.nii.gz
  Size: 242316193 	Blocks: 473288     IO Block: 4096   regular file
Device: 802h/2050d	Inode: 2361085     Links: 2
Access: (0755/-rwxr-xr-x)  Uid: ( 1001/nguenther)   Gid: ( 1001/nguenther)
Access: 2021-05-15 15:03:15.213585915 -0400
Modify: 2021-05-15 13:59:55.595508952 -0400
Change: 2021-05-15 14:16:35.929312611 -0400
 Birth: 2021-05-15 13:59:53.271495454 -0400
```

And you can find the two hardlinks with this:

```
nguenther@data:~/datasets/msseg_challenge_2016$ find -inum "$(stat -c '%i' sub-01016SACH/anat/sub-01016SACH_ce-T1w.nii.gz)"
./sub-01016SACH/anat/sub-01016SACH_ce-T1w.nii.gz
./.git/annex/objects/xp/9m/SHA256E-s242316193--a715d9de115cc083dab964ceaad84fe438428509bd33e155b6b4c518f983a6ef.nii.gz/SHA256E-s242316193--a715d9de115cc083dab964ceaad84fe438428509bd33e155b6b4c518f983a6ef.nii.gz
```

(I think the OS X slang of that command is `find . -inum "$(stat -F '%i' sub-01016SACH/anat/sub-01016SACH_ce-T1w.nii.gz)"`)

**Comment by jcohenadad** [comment]
> I'm not sure what the problem with the smb://duke copy is; I see only one time the size:

small precision: I am using AFP.

**Comment by jcohenadad** [comment]
> There's another way to check: pick a file and run stat; you'll see it has "Links: 2":

hum, looks like the `stat` on macOs is different: 
```console
julien-macbook:~/data.neuro/msseg_challenge_2021 $ stat sub-013/ses-01/anat/sub-013_ses-01_acq-middlespace_FLAIR.nii.gz 
16777220 98410205 -rwxr-xr-x 2 julien staff 0 35804487 "May 27 22:19:13 2021" "May 27 15:21:37 2021" "May 27 15:21:37 2021" "May 27 15:21:32 2021" 4096 70224 0 sub-013/ses-01/anat/sub-013_ses-01_acq-middlespace_FLAIR.nii.gz
```

```console
julien-macbook:~/data.neuro/msseg_challenge_2021 $ find -inum "$(stat -c '%i' sub-01016SACH/anat/sub-01016SACH_ce-T1w.nii.gz)"
stat: illegal option -- c
usage: stat [-FlLnqrsx] [-f format] [-t timefmt] [file ...]
find: illegal option -- i
usage: find [-H | -L | -P] [-EXdsx] [-f path] path ... [expression]
       find [-H | -L | -P] [-EXdsx] -f path [path ...] [expression]
```

**Comment by kousu** [comment]
Hey @jcohenadad please try this:

> (I think the OS X slang of that command is find . -inum "$(stat -f '%i' sub-01016SACH/anat/sub-01016SACH_ce-T1w.nii.gz)")

`98410205` is the inode pointer.

**Comment by kousu** [comment]
And using AFP is probably the source of the weirdness you saw with duke. AFP probably doesn't transit hardlinks. I'm a little shocked SMB does!

**Comment by hermancollin** [comment]
@jcohenadad it could actually be simpler than this for the doubling problem

the `du` command shows disk usage in block size (which is related to the filesystem used) when `-h` option is *not* specified. My guess is that your device's blocksize is 512 (or half a KB). this could explain why `du -s` gives 43640816 half kilobytes, or 21820408 kilobytes which gives 21G with `-h` flag

**Comment by joshuacwnewton** [comment]
> My guess is that your device's blocksize is 512 (or half a KB). this could explain why `du -s` gives 43640816 half kilobytes, or 21820408 kilobytes (21G with `-h` flag)

This sounds like the most reasonable explanation to me. Also, to verify, you could compare `du -sB 1024` and `du -sh` (to explicitly set a blocksize of 1KB).

**Comment by jcohenadad** [comment]
> Also, to verify, you could compare du -sh with du -sB 1024

Unfortunately "-B" does not work in macOs:
~~~
ulien-macbook:~/data.neuro $ du -sB 1024 msseg_challenge_2021/
du: illegal option -- B
usage: du [-H | -L | -P] [-a | -s | -d depth] [-c] [-h | -k | -m | -g] [-x] [-I mask] [file ...]
~~~

Thank you for your insights @hermancollin ! It looks like both my local git-annexed and the one on duke (ie: "true" size) have the same size, meaning that the `annex.thin true` trick worked 🎉:
~~~
julien-macbook:~/data.neuro $ du -sh msseg_challenge_2021/
5.5G	msseg_challenge_2021/
julien-macbook:~/data.neuro $ du -sh /Volumes/projects/ivadomed/ms_challenge_2021_bids_DoNotUse 
5.3G	/Volumes/projects/ivadomed/ms_challenge_2021_bids_DoNotUse
~~~

But then, I'm not sure why my app is giving me 11.8 GB:
![image](https://user-images.githubusercontent.com/2482071/120018241-e57b1780-bfb4-11eb-88c9-22c8336b172c.png)

Is it wrong, or should "GB" be interpreted as blocksize*10^9

**Comment by jcohenadad** [comment]
Update on the graph above: what's strange is that both the .git and sub-* folders give me "data of non-negligible size":
![Screen Shot 2021-05-28 at 1 03 18 PM](https://user-images.githubusercontent.com/2482071/120018475-3428b180-bfb5-11eb-9f9b-3693ccdb057d.png)
![Screen Shot 2021-05-28 at 1 03 07 PM](https://user-images.githubusercontent.com/2482071/120018479-3559de80-bfb5-11eb-9d9d-08e0060f3d1d.png)

So maybe there is a problem with hardlinks there?

**Comment by kousu** [comment]
I think your app doesn't understand hardlinks. It's counting each folder by summing up the filesizes. To deal with hardlinks to compute actual disk space used you need to keep a table of inode pointers.

What does Finder say about it?

**Comment by kousu** [comment]
A file is an array of bytes scattered over a disk.
An inode is the metadata containing the pointers to the bytes.
An inode *number* is the pointer to the inode itself.
A file on disk is a pathname that points to the inode *number*.

A **hardlink** is simply a second file with the same pointer in it; every file that points to the same inode *number* is the same file.

**Comment by kousu** [comment]
And this is what `annex.thin` enables (though it's very poorly documented IMO): it creates hardlinks *between the checkout folder ./* and the files in .git/annex/.

**Comment by joshuacwnewton** [comment]
> Unfortunately "-B" does not work in macOs:

Ah, I see. It looks like you want to use `-k` then on macOS. From the [manpage](https://ss64.com/osx/du.html):

```
    -k    KB Display block counts in 1024-byte (1-Kbyte) blocks.
```

**Comment by jcohenadad** [comment]
> What does Finder say about it? 

Ah! Finder says 5.68GB, so consistent with `du` 👇
![Screen Shot 2021-05-28 at 2 54 10 PM](https://user-images.githubusercontent.com/2482071/120029890-b5d40b80-bfc4-11eb-9b54-622c5f33bc93.png)

> Or `df`?

I don't know how to use `df` for a folder only (and no time to try to remove/add it back).

> Ah, I see. It looks like you want to use du -sk then on macOS. From the manual page:

ah! thanks, here you go:
~~~
julien-macbook:~/data.neuro $ du -sk msseg_challenge_2021/
5775552	msseg_challenge_2021/
~~~

thank you all for your insights! TIL

**Comment by joshuacwnewton** [comment]
Thanks for asking all of your questions! I had no idea `du` behaved differently between OSs. We're all learning. :smile:

**Comment by kousu** [comment]
> I don't know how to use `df` for a folder only (and no time to try to remove/add it back).

Indeed. Try it next time if you're still unsure. do `df .; git annex get; df .`.

**Comment by kousu** [comment]
Back to the original question, here's a better way @alexfoias thought up with me to track dataset size: write a netdata collector to run `du` on each git repo like I prototyped for `smb://duke` (https://github.com/neuropoly/computers/issues/8#issuecomment-795728672). The UI then is just looking at https://monitor.neuro.polymtl.ca/host/data.neuro.polymtl.ca/ -- there will be a set of charts in there showing.

I could give each repo its own chart, or we could have one chart with multiple lines.

**Comment by kousu** [comment]
Only tangentially related but interesting: Github, by comparison, *also can't tell you how big a repo is*. Their API [includes a "size" field](https://docs.github.com/en/rest/reference/repos#get-a-repository-readme) but it's wrong because:

There seem to be bugs? https://stackoverflow.com/questions/8646517/how-can-i-see-the-size-of-a-github-repository-before-cloning-it

And a 'fork' means `git clone -s`, which creates .git/objects/info/alternates, a kind of pseudo-symlink, which means that `du -hs` *fails* to count the actual size of the repo (https://git-blame.blogspot.com/2012/08/bringing-bit-more-sanity-to-alternates.html)

Plus:

> Here's a hint: My absolute largest repository contains only images of various formats, it's an "artwork" repo of icons which I use in various apps. Yet, GitHub reports the size as 0. So I'm assuming it only considers the size of known source files, and doesn't consider unknown file types. – [Jerry Dodge Jun 14 '19 at 19:27](https://stackoverflow.com/questions/8646517/how-can-i-see-the-size-of-a-github-repository-before-cloning-it#comment99784402_8646517)

**Comment by bpinsard** [comment]
Late to the party, crashed here looking for others stuffs. 
`git-annex info` can get you the size of the annexed keys on the different remotes where the files are stored (here, or special or normal remotes), provided the annex branch is up-to-date.


``` 
git-annex info here
uuid: 2f59cb00-8b03-4884-995a-baa285129730
description: basile@blob-XPS-9320:~/data/tests/ds000001 [here]
trust: semitrusted
remote annex keys: 80
remote annex size: 2.42 gigabytes
``` 

```
git-annex info s3-PUBLIC
uuid: 8d2b6e96-ad81-44a5-99b4-0ec37d6b3800
...
...
remote annex keys: 141
remote annex size: 2.42 gigabytes (+ 61 unknown size)
``` 

As for git, when you only need the latest state of the dataset, shallow clone can be more efficient, notably for short lived clones (eg github actions runner).
